chrome/content/zotero/zoteroPane.js

```diff
@@ -3080,6 +3090,52 @@ var ZoteroPane = new function()
 		// add locate menu options
 		yield Zotero_LocateMenu.buildContextMenu(menu, true);
 	});
+
+
+	this.buildAddToCollectionMenu = function (event) {
+		if (event.target.id !== 'zotero-add-to-collection-popup') return;
+
+		let popup = document.querySelector('#zotero-add-to-collection-popup');
+		let separator = popup.querySelector('#zotero-add-to-collection-separator');
+		while (popup.childElementCount > 2) {
+			popup.removeChild(popup.firstElementChild);
+		}
+
+		let items = Zotero.Items.keepParents(this.getSelectedItems());
+		let collections = Zotero.Collections.getByLibrary(this.getSelectedLibraryID());
+		for (let col of collections) {
+			let menuItem = Zotero.Utilities.Internal.createMenuForTarget(
+				col,
+				popup,
+				null,
+				(event, collection) => {
+					if (event.target.tagName == 'menuitem') {
+						this.addItemsToCollection(collection);
+						event.stopPropagation();
+					}
+				},
+				collection => items.every(item => collection.hasItem(item))
+			);
+			separator.before(menuItem);
+		}
+
+		separator.setAttribute('hidden', !collections.length);
+	};
+
+
+	this.addItemsToCollection = async function (collection, createNew = false) {
+		// Get items first because newCollection() will deselect
+		let items = Zotero.Items.keepParents(this.getSelectedItems());
+
+		if (createNew) {
```

Comment:
Throw error (e.g., "collection must be null if createNew is true") if `collection` is not falsy